### PR TITLE
Basic rest api with user endpoints and auth

### PR DESCRIPTION
**Pull request by rossem**
- auth
- some interface functions with user model
- user get/post
- tests for api
- use test db for testing

### CODE REVIEW

**Comment by rossem** [comment]
closes #4

File: core/router.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+	"github.com/julienschmidt/httprouter"
+)
+
+func InitRouter() *httprouter.Router {
+	router := httprouter.New()
+
+	router.GET("/api/v1/", index)
+	router.GET("/api/v1/users/:userId", auth(getUser))
+	router.POST("/api/v1/users", createUser)
```
Comment (Wojtechnology) [inline core/router.go]: We don't need the v1 for v1. We'll start at v2

File: core/router.go [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+	return router
+}
+
+func index(w http.ResponseWriter, r *http.Request, _ httprouter.Params) {
+	fmt.Fprintf(w, "le iterable api")
```
Comment (Wojtechnology) [inline core/router.go]: Actually write something to the response instead of logging.

File: db/db.go [hunk 3]
```diff
@@ -24,3 +25,11 @@ func buildDataSourceName() string {
...
 		config.Global.Postgres.SSLMode)
 }
+
+func init() {
+	log.Println("Starting DB")
+	if err := ConnectSQL(); err != nil {
+		log.Fatalf("Failed to connect to DB: %v", err)
+	}
+	log.Println("DB started")
+}
```
Comment (Wojtechnology) [inline db/db.go:35]: Move back to main.go plz

File: db/user.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+var ErrNoUser = errors.New("no user")
+
+type User struct {
+	Id    int    `json:"id"`
+	Email string `json:"email"`
+	Token string `json:"token"`
+}
+
+func FindUserBy(column string, value interface{}) (*User, error) {
```
Comment (Wojtechnology) [inline db/user.go]: I feel like you should abstract this whole thing into db.go. As in make a generic function for getting x model by y attribute.

File: core/auth.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+			return
+		}
+		setCurrentUser(r, user)
+		h(w, r, ps)
+		context.Clear(r)
+	}
+}
+
+func getCurrentUser(r *http.Request) *db.User {
+	return context.Get(r, "user").(*db.User)
```
Comment (Wojtechnology) [inline core/auth.go:36]: What does this do?

File: db/user.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+	u := &User{}
+	q := fmt.Sprintf("SELECT id, email, token FROM users WHERE %v=$1", column)
```
Comment (Wojtechnology) [inline db/user.go]: Abstract the queries out into very generic methods in db.go

File: db/user.go [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+	return FindUserBy("token", token)
+}
+
+func CreateUser(email string) (*User, error) {
+	token, err := generateUniqueToken()
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = db.Exec(`INSERT INTO users (email, token) VALUES ($1, $2)`, email, token)
```
Comment (Wojtechnology) [inline db/user.go]: Also here.

File: db/user.go [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+	if err != nil {
+		return nil, err
```
Comment (Wojtechnology) [inline db/user.go]: Can we get a better idea of what the error is and respond with an appropriate message?

**Comment by Wojtechnology** [comment]
Do some stuff then merge

File: db/user.go [hunk 9]
```diff
@@ -0,0 +1,81 @@
+package db
```
Comment (narensrini) [inline db/user.go:1]: I thought we were having models in a models directory

Comment (rossem) [inline db/user.go:1]: then I'd have to do `db.DB`..

File: config/config_test.go [hunk 10]
```diff
@@ -18,10 +18,11 @@ func TestLoad(t *testing.T) {
 
 	conf := IterableConfig{}
 	err := load(&conf, "test.yaml", yaml.Unmarshal)
-
 	assert.Nil(err, "There was no error")
-	assert.Equal(conf.Postgres.DBName, "iterable_test", "This should be the same as test.yaml")
-	assert.Equal(conf.Postgres.SSLMode, "disable", "This should be the same as test.yaml")
+
+	same := "This should be the same as test.yaml"
```
Comment (narensrini) [inline config/config_test.go:23]: `same`

File: core/auth.go (hunk 5 above)
Comment (rossem) [inline core/auth.go:36]: Like this function? It just gets the current user from the request context

**Comment by rossem** [comment]
alright, merging this. we can implement fixes in separate pr's afterwards. this one is big enough already.

**Pull request merged**